src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,344 @@
+            }
+
+            return TryParseInternal(s.AsSpan(), out result, out _);
+        }
+
+        public static bool TryParse(ReadOnlySpan<char> s, out IPNetwork result)
+        {
+            return TryParseInternal(s, out result, out _);
+        }
+        #endregion
+
+        #region Private methods
+        private static bool TryParseInternal(ReadOnlySpan<char> s, out IPNetwork result, [NotNullWhen(false)] out string? error)
+        {
+            const int maxCharsCountAfterIpAddress = 4;
+            const int minCharsRequired = 4;
+
+            if (s.Length < minCharsRequired)
+            {
+                error = SR.dns_bad_ip_network;
+                result = default;
+                return false;
+            }
+
+            int separatorExpectedStartingIndex = s.Length - maxCharsCountAfterIpAddress;
+            int separatorIndex = s
+                .Slice(separatorExpectedStartingIndex)
+                .IndexOf(addressAndPrefixLengthSeparator);
+            if (separatorIndex != -1)
+            {
+                separatorIndex += separatorExpectedStartingIndex;
+
+                var ipAddressSpan = s.Slice(0, separatorIndex);
+                var prefixLengthSpan = s.Slice(separatorIndex + 1);
+
+                if (IPAddress.TryParse(ipAddressSpan, out var ipAddress) && int.TryParse(prefixLengthSpan, out var prefixLength))
+                {
+                    result = new IPNetwork
+                    {
+                        BaseAddress = ipAddress,
+                        PrefixLength = prefixLength
+                    };
+
+                    error = result.Validate()?.ParseExceptionMessage;
+                    return error == null;
+                }
+                else
+                {
+                    error = SR.dns_bad_ip_network;
+                    result = default;
+                    return false;
+                }
+            }
+            else
+            {
+                error = SR.dns_bad_ip_network;
+                result = default;
+                return false;
+            }
+        }
+
+        private readonly record struct ValidationError(
+            Func<Exception> ConstructorExceptionFactoryMethod,
+            string ParseExceptionMessage);
+
+        private static readonly ValidationError
+            _baseAddressIsNullError = new ValidationError(() => new ArgumentNullException(baseAddressConstructorParamName), string.Empty),
+            _baseAddressHasSetBitsInMaskError = new ValidationError(() => new ArgumentException(baseAddressConstructorParamName), SR.dns_bad_ip_network),
+            _prefixLengthLessThanZeroError = new ValidationError(() => new ArgumentOutOfRangeException(prefixLengthConstructorParamName), SR.dns_bad_ip_network),
+            _prefixLengthGreaterThanAllowedError = new ValidationError(() => new ArgumentOutOfRangeException(prefixLengthConstructorParamName), SR.dns_bad_ip_network);
+
+        private ValidationError? Validate()
```

Comment:
My recommendation to implement reusable validation logic would be to do something like this:

```C#
private static bool Validate(IPAddress baseAddress, int prefixLength, out Func<Exception>? exceptionFactory)
{
    // It's sufficient to do the baseAddress == null validation in the constructor.

    if (prefixLength < 0 || prefixLength > GetAddressFamilyByteLength(baseAddress.AddressFamily) * bitsPerByte)
    {
        exceptionFactory = static () => new ArgumentOutOfRangeException(nameof(prefixLength));
    }

    // any further validation

    exceptionFactory = null;
    return true;
}
```

This way you could avoid defining constants, and storing exceptions. When using it in the public `Parse(...)`, you can make the resulting validation exception as the inner exception of the `FormatException` being thrown there.